src/circuit/gadget/ecc/chip/mul.rs

```diff
@@ -0,0 +1,427 @@
+    fn create_init_scalar_gate(&self, meta: &mut ConstraintSystem<pallas::Base>) {
+        meta.create_gate("Initialize running sum for variable-base mul", |meta| {
+            let q_init_z = meta.query_selector(self.q_init_z);
+            let z = meta.query_advice(self.hi_config.z, Rotation::cur());
+
+            vec![q_init_z * z]
+        });
+    }
+
+    /// Gate used to check final scalar is recovered.
+    fn create_final_scalar_gate(&self, meta: &mut ConstraintSystem<pallas::Base>) {
+        meta.create_gate("Decompose scalar for variable-base mul", |meta| {
+            let q_mul_decompose_var = meta.query_selector(self.q_mul_decompose_var);
+            let scalar = meta.query_advice(self.scalar, Rotation::cur());
+            let z_cur = meta.query_advice(self.z_complete, Rotation::cur());
+
+            // The scalar field `F_q = 2^254 + t_q`.
+            // -((2^127)^2) = -(2^254) = t_q (mod q)
+            let t_q = -(pallas::Scalar::from_u128(1u128 << 127).square());
+            let t_q = pallas::Base::from_bytes(&t_q.to_bytes()).unwrap();
+
+            // Check that `k = scalar + t_q`
+            vec![q_mul_decompose_var * (scalar + Expression::Constant(t_q) - z_cur)]
+        });
+    }
+
+    pub(super) fn assign_region(
+        &self,
+        scalar: &CellValue<pallas::Base>,
+        base: &EccPoint,
+        offset: usize,
+        region: &mut Region<'_, pallas::Base>,
+    ) -> Result<EccPoint, Error> {
+        // Decompose the scalar bitwise (big-endian bit order).
+        let bits = decompose_for_scalar_mul(scalar.value());
+        let bits_incomplete_hi = &bits[INCOMPLETE_HI_RANGE];
+        let bits_incomplete_lo = &bits[INCOMPLETE_LO_RANGE];
+        let lsb = bits[pallas::Scalar::NUM_BITS as usize - 1];
+
+        // Initialize the accumulator `acc = [2]base`
+        let acc = self
+            .add_config
+            .assign_region(&base, &base, offset, region)?;
+
+        // Increase the offset by 1 after complete addition.
+        let offset = offset + 1;
+
+        // Initialize the running sum for scalar decomposition to zero
+        let z_init = {
+            // Constrain the initialization of `z` to equal zero.
+            self.q_init_z.enable(region, offset)?;
+
+            let z_val = pallas::Base::zero();
+            let z_cell =
+                region.assign_advice(|| "initial z", self.hi_config.z, offset, || Ok(z_val))?;
+
+            Z(CellValue::new(z_cell, Some(z_val)))
+        };
+
+        // Increase the offset by 1 after initializing `z`.
+        let offset = offset + 1;
+
+        // Double-and-add (incomplete addition) for the `hi` half of the scalar decomposition
+        let (x_a, y_a, zs_incomplete_hi) = self.hi_config.double_and_add(
+            region,
+            offset,
+            &base,
+            bits_incomplete_hi,
+            (X(acc.x), Y(acc.y.value()), z_init),
```

Comment:
We don't use `acc.y.cell()` anywhere, so we aren't constraining the output of the initial complete addition to be the input to the incomplete addition. We need to pass `acc.y` in entirely, and then inside `double_and_add` have the initial values of lambda_1 etc. be appropriately constrained.